crates/accelerate/src/convert_2q_block_matrix.rs

```diff
@@ -32,32 +37,61 @@ pub fn blocks_to_matrix(
     py: Python,
     op_list: Vec<(PyReadonlyArray2<Complex64>, SmallVec<[u8; 2]>)>,
 ) -> PyResult<Py<PyArray2<Complex64>>> {
-    let identity = aview2(&ONE_QUBIT_IDENTITY);
-    let input_matrix = op_list[0].0.as_array();
-    let mut matrix: Array2<Complex64> = match op_list[0].1.as_slice() {
-        [0] => kron(&identity, &input_matrix),
-        [1] => kron(&input_matrix, &identity),
+    let identity = aview2(&ONE_QUBIT_IDENTITY).into_faer_complex();
+    let input_matrix = op_list[0].0.as_array().into_faer_complex();
+
+    let mut matrix = match op_list[0].1.as_slice() {
+        [0] => identity.kron(input_matrix.as_ref()),
+        [1] => input_matrix.kron(identity.as_ref()),
         [0, 1] => input_matrix.to_owned(),
-        [1, 0] => change_basis(input_matrix),
-        [] => Array2::eye(4),
+        [1, 0] => change_basis_faer(input_matrix),
+        [] => Mat::<c64>::identity(4, 4),
         _ => unreachable!(),
     };
     for (op_matrix, q_list) in op_list.into_iter().skip(1) {
-        let op_matrix = op_matrix.as_array();
+        let op_matrix = op_matrix.as_array().into_faer_complex();
 
         let result = match q_list.as_slice() {
-            [0] => Some(kron(&identity, &op_matrix)),
-            [1] => Some(kron(&op_matrix, &identity)),
-            [1, 0] => Some(change_basis(op_matrix)),
-            [] => Some(Array2::eye(4)),
-            _ => None,
-        };
-        matrix = match result {
-            Some(result) => result.dot(&matrix),
-            None => op_matrix.dot(&matrix),
+            [0] => identity.kron(op_matrix.as_ref()),
+            [1] => op_matrix.kron(identity.as_ref()),
+            [1, 0] => change_basis_faer(op_matrix),
+            [] => Mat::<c64>::identity(4, 4),
+            _ => op_matrix.to_owned(),
```

Comment:
The `op_matrix.to_owned()` here is a bit weird to me. It feels like an unnecessary copy as we don't actually need an owned copy of the array from python here as it's only used in the `matmul` call below. The previous logic was a bit ugly but was done that way to avoid copying with `to_owned()`. The other option would be to maybe try something like:

```rust
            [0] => identity.kron(op_matrix.as_ref()).as_ref(),
            [1] => op_matrix.kron(identity.as_ref()).as_ref(),
            [1, 0] => change_basis_faer(op_matrix).as_ref(),
            [] => Mat::<c64>::identity(4, 4).as_ref(),
            _ => op_matrix.to_owned(),
```

and change `result` to be a `MatRef<c64>` instead of of a `Mat<c64>`. Although I'm not sure if the compiler would be happy with this (there might be a lifetime issue doing this. The other advantage is if `result` becomes a `MatRef` we can create the 4x4 identity matrix as a static and avoid the allocation (although in practice I'm not sure this ever gets used so it probably doesn't matter).

Reply:
I have tried the code you suggested, and as you said, the compiler is not happy about it. The error message is "temporary value is freed at the end of this statement" referring to each case of the `match` where we use `as_ref()`.

Eric and I have been trying to fix it, but the only alternative we saw to avoid copying with `to_owned()` is to recover the previous logic, making the `result` matrix an `Option<Mat<c64>>` and "duplicating" the `matmul` call using the unwrapped value of the `Some` or `op_matrix` in case of `None`. It's not as readable as now, but performance-wise should be better.

Here is the `git diff` of the change. What do you think about doing something like this again? I suspect we will see this case in `two_qubit_decompose.rs` too.

```diff
         let op_matrix = op_matrix.as_array().into_faer_complex();
 
         let result = match q_list.as_slice() {
-            [0] => identity.kron(op_matrix.as_ref()),
-            [1] => op_matrix.kron(identity.as_ref()),
-            [1, 0] => change_basis_faer(op_matrix),
-            [] => Mat::<c64>::identity(4, 4),
-            _ => op_matrix.to_owned(),
+            [0] => Some(identity.kron(op_matrix.as_ref())),
+            [1] => Some(op_matrix.kron(identity.as_ref())),
+            [1, 0] => Some(change_basis_faer(op_matrix)),
+            [] => Some(Mat::<c64>::identity(4, 4)),
+            _ => None,
         };
 
         let aux = matrix.clone();
-        matmul(
-            matrix.as_mut(),
-            result.as_ref(),
-            aux.as_ref(),
-            None,
-            c64::new(1., 0.),
-            Parallelism::None,
-        );
+        match result {
+            Some(x) => {
+                matmul(
+                    matrix.as_mut(),
+                    x.as_ref(),
+                    aux.as_ref(),
+                    None,
+                    c64::new(1., 0.),
+                    Parallelism::None,
+                );
+            },
+            None => {
+                matmul(
+                    matrix.as_mut(),
+                    op_matrix,
+                    aux.as_ref(),
+                    None,
+                    c64::new(1., 0.),
+                    Parallelism::None,
+                );
+            }
+        };
     }

```

Reply:
FYI this is what we were trying to do:

```rust
let result = match q_list.as_slice() {
  [0] => Some(identity.kron(op_matrix.as_ref())),
  [1] => Some(op_matrix.kron(identity.as_ref())),
  [1, 0] => Some(change_basis_faer(op_matrix)),
  [] => Some(Mat::<c64>::identity(4, 4)),
   _ => None,
};
let arg1 = match result {
   // Go from Mat to MatRef. This failed because the lifetime of x
  Some(x) => x.as_ref(),
  // op_matrix is already MatRef
  None => op_matrix
}
```

Using `result.as_ref().unwrap_or(op_matrix)` does not work. `result.as_ref` results in an `Option<&Mat>` rather than `Option<MatRef>` and they are not the same thing.

Reply:
Yeah, I think lets use the pattern it's ugly but it lets us avoid a copy

Reply:
> Using `result.as_ref().unwrap_op(op_matrix)` does not work.

Should this not be `result.as_ref().map(|x| x.as_ref()).unwrap_op(op_matrix)`?

Reply:
(see https://github.com/Qiskit/qiskit/pull/12193#discussion_r1579360316)

Reply:
> Should this not be result.as_ref().map(|x| x.as_ref()).unwrap_op(op_matrix)?

Yes, that works! Much better than duplicating the `matmul` call, thanks Jake!